### PR TITLE
rpc: Add getrpcinfo command

### PR DESCRIPTION
**Pull request by promag**
The new `getrpcinfo` command exposes details of the RPC interface. The details can be configuration properties or runtime values/stats.

This can be particular useful to coordinate concurrent functional tests (see #14958 from where this was extracted).

### CODE REVIEW

File: src/rpc/server.cpp [hunk 1]
```diff
@@ -31,12 +31,12 @@ static std::string rpcWarmupStatus GUARDED_BY(cs_rpcWarmup) = "RPC server starte
 static RPCTimerInterface* timerInterface = nullptr;
 /* Map of name to timer. */
 static std::map<std::string, std::unique_ptr<RPCTimerBase> > deadlineTimers;
+static std::shared_ptr<int> g_active_commands_counter = std::make_shared<int>(0);
```
Comment (promag) [inline src/rpc/server.cpp]: This is temporary, just for concept ACK.

File: src/rpc/server.cpp [hunk 2]
```diff
@@ -255,11 +255,28 @@ static UniValue uptime(const JSONRPCRequest& jsonRequest)
...
+        throw std::runtime_error(
+            RPCHelpMan{"getrpcino",
+                "\nReturns details of the RPC server.\n", {}}
+                .ToString()
+        );
+    }
+
+    UniValue result(UniValue::VOBJ);
+    result.pushKV("active_commands", (uint64_t) g_active_commands_counter.use_count() - 1);
+    result.pushKV("uptime", GetTime() - GetStartupTime());
```
Comment (promag) [inline src/rpc/server.cpp]: This is an example.

File: src/rpc/server.cpp [hunk 3]
```diff
@@ -255,11 +255,28 @@ static UniValue uptime(const JSONRPCRequest& jsonRequest)
...
+    return result;
+}
+
 // clang-format off
 static const CRPCCommand vRPCCommands[] =
 { //  category              name                      actor (function)         argNames
   //  --------------------- ------------------------  -----------------------  ----------
     /* Overall control/query calls */
+    { "control",            "getrpcinfo",          &getrpcinfo,          {}  },
```
Comment (Empact) [inline src/rpc/server.cpp]: nit: whitespace

File: src/rpc/server.cpp [hunk 4]
```diff
@@ -255,11 +255,28 @@ static UniValue uptime(const JSONRPCRequest& jsonRequest)
...
+    if (request.fHelp || request.params.size() > 0) {
+        throw std::runtime_error(
+            RPCHelpMan{"getrpcino",
+                "\nReturns details of the RPC server.\n", {}}
+                .ToString()
+        );
+    }
+
+    UniValue result(UniValue::VOBJ);
+    result.pushKV("active_commands", (uint64_t) g_active_commands_counter.use_count() - 1);
```
Comment (Empact) [inline src/rpc/server.cpp]: `"active_command_count"`?

File: src/rpc/server.cpp [hunk 5]
```diff
@@ -255,11 +255,28 @@ static UniValue uptime(const JSONRPCRequest& jsonRequest)
     return GetTime() - GetStartupTime();
 }
 
+static UniValue getrpcinfo(const JSONRPCRequest& request)
+{
+    if (request.fHelp || request.params.size() > 0) {
+        throw std::runtime_error(
+            RPCHelpMan{"getrpcino",
```
Comment (practicalswift) [inline src/rpc/server.cpp]: Should be "getrpcinfo"? :-)

File: src/rpc/server.cpp [hunk 6]
```diff
@@ -484,10 +501,9 @@ UniValue CRPCTable::execute(const JSONRPCRequest &request) const
     if (!pcmd)
         throw JSONRPCError(RPC_METHOD_NOT_FOUND, "Method not found");
 
-    g_rpcSignals.PreCommand(*pcmd);
-
     try
     {
+        auto active_command = g_active_commands_counter;
```
Comment (practicalswift) [inline src/rpc/server.cpp]: `active_command` unused? :-)

**Comment by promag** [comment]
@Empact @practicalswift I guess you concept ACK?

File: src/rpc/server.cpp (hunk 6 above)
Comment (promag) [reply src/rpc/server.cpp]: Ya, the idea is to have a RAII counter and for now I'm using the shared ptr usage counter 😅

**Comment by ch4ot1c** [comment]
Concept Ack.

Does `uptime` belong in the response? Should it be a separate `rpc_uptime`, measured instead from `StartRPC` / the `Started` signal?: https://github.com/bitcoin/bitcoin/blob/24b3b788be9bdf24019c5e46a7943db7c4328389/src/init.cpp#L733

File: src/rpc/server.cpp (hunk 1 above)
Comment (jonasschnelli) [reply src/rpc/server.cpp]: Yeah. I suggest using a manual increment rather then the `use_count()`

**Comment by jonasschnelli** [comment]
Concept ACK.
* What about showing active commands (things like `rescanblockchain` and `scantxoutset`)?

**Comment by practicalswift** [comment]
Concept ACK

File: src/rpc/server.cpp (hunk 5 above)
Comment (instagibbs) [reply src/rpc/server.cpp]: "getalpcino"

**Comment by instagibbs** [comment]
concept ACK . could it add things like port it's listening to?

**Comment by promag** [comment]
I guess you have to know that to call this RPC?

**Comment by instagibbs** [comment]
@promag the computer knows this, yes ;P

File: src/rpc/server.cpp (hunk 5 above)
Comment (promag) [reply src/rpc/server.cpp]: @instagibbs 🤣 @MarcoFalke what?

**Comment by laanwj** [comment]
So I'd say if everyone is Concept ACK, I think it makes sense to fix the nits, utACK and merge this and add further ideas in later PRs.

**Comment by promag** [comment]
@laanwj updated.

**Comment by laanwj** [comment]
utACK a0ac15459a0df598e1ee1fd36a3899a129cecaeb

**Pull request merged**